jaxopt/_src/tree_util.py

```diff
@@ -102,4 +108,40 @@ def tree_gram(a):
   """
   vmap_left = jax.vmap(tree_vdot, in_axes=(0,None))
   vmap_right = jax.vmap(vmap_left, in_axes=(None,0))
-  return vmap_right(a, a)
\ No newline at end of file
+  return vmap_right(a, a)
+
+
+def tree_inf_norm(v):
+  """Compute infinity norm of pytree."""
+  leaf_inf_norm = tree_map(lambda l: jnp.max(jnp.abs(l)), v)
+  return tree_reduce(jnp.maximum, leaf_inf_norm)
```

Comment:
Since I assume `tree_inf_norm` returns a scalar value, I think the correct operator should be `jnp.max`.

Reply:
`leaf_inf_norm` is a pytree of scalars, so to perform reduction we need an associative binary operation (here `jnp.maximum`). The operator `jnp.max` would work on ndarray, but does not support pytrees. So to apply `jnp.max` we would need to concatenate leaves of `leaf_inf_norm` first, which would not be as efficient.

Reply:
Ok, thanks for the clarification.